app/models/mentorship_matcher.rb

```diff
@@ -0,0 +1,106 @@
+class MentorshipMatcher
+  LANGUAGE_MATCH_SCORE = 40
+  LOCAL_PREFERENCE_SCORE = 15
+  REMOTE_PREFERENCE_SCORE = 5
+  DISTANCE_SCORES = {
+    (0..300) => 30,        # Same timezone
+    (301..1000) => 10      # Near timezone (+/- 1-2 hours)
+  }.freeze
+  LONG_DISTANCE_SCORE = 5  # Distant timezone
+
+  def initialize(applicant)
+    @applicant = applicant
+    @applicant_languages = applicant.languages
+    @applicant_questionnaire = applicant.applicant_questionnaire
+  end
+
+  def matches
+    scored_mentors = available_mentors.map do |mentor|
+      [mentor, calculate_match_score(mentor)]
+    end
+
+    scored_mentors.sort_by { |_, score| -score }
+  end
+
+  private
+
+  attr_reader :applicant, :applicant_languages, :applicant_questionnaire
+
+  def available_mentors
+    Mentorship.available_mentors.includes(:languages, :mentor_questionnaire)
+  end
+
+  # Calculates match score (max 100 points) based on:
+  # 1. Languages: 40 points for any shared language
+  # 2. Location: 30 points same timezone, 10 points near timezone (+/-2h), 5 points distant timezone
+  # 3. Preferences: Each matching preference (code/career)
+  #    - 15 points if in same/near timezone
+  #    - 5 points if in distant timezone
```

Comment:
Really like this.

Think we should _add_ country. i.e. if the same country 40 points. (In place of language?)